pageserver/src/tenant/timeline.rs

```diff
@@ -3104,6 +3105,24 @@ impl Timeline {
 
         Ok(image_layers)
     }
+
+    /// Wait until the background initial logical size calculation is complete, or
+    /// this Timeline is shut down.  Calling this function will cause the initial
+    /// logical size calculation to skip waiting for the background jobs barrier.
+    pub(crate) async fn await_initial_logical_size(self: Arc<Self>) {
+        if let Some(await_bg_cancel) = self
+            .current_logical_size
+            .cancel_wait_for_background_loop_concurrency_limit_semaphore
+            .get()
+        {
+            await_bg_cancel.cancel();
+        }
+
+        tokio::select!(
+            _ = self.current_logical_size.initialized.notified() => {},
+            _ = self.cancel.cancelled() => {}
+        )
+    }
```

Comment:
`cancel_wait_for_background_loop_concurrency_limit_semaphore` is always `Some()` because we call this function after activating the timeline.
IMO it's an error to call this function if it's still `None`.
WHy is it an error? Because we move on waiting for the size to be calculated.

In https://github.com/neondatabase/neon/pull/6010 I introduce `get_current_logical_size_exact_wait`.
I think we should use that function here.

Alternatively, we should return an error from this function if the `cancel_wait_for_background_loop_concurrency_limit_semaphore` is `None`, so that our caller can skip waiting for initial logical size to become available - if they didn't expedite the calculation of initial logical size because they couldn't `await_bg_cancel.cancel();`, they shouldn't be waiting for it to become available any time soon.

Reply:
I thought it was the opposite; if `cancel_wait_for_background_loop_concurrency_limit_semaphore` is None, that indicates that someone else already set the cancellation token, so it should proceed.

However, the priority should be to avoid deadlocks, so I'll defensively make this function drop out if it can't get the cancel token, and log 8c7bccfba -- with a TODO in there that this code path is getting quite twisty.  Let's revisit next week and/or when 6010 hits.